syntax = "proto3";

package component_db.v1;

/**
  * サービスインターフェース情報を設定/取得するためのパラメータ
  */
message ServiceInterfaceInfo {
  string name = 1;				// インターフェース名
  uint32 direction = 2;			// 方向(1:Provided, 2:Required)
  string interface_type = 3;	// インターフェース型
}

/**
  * サービスポート情報を設定/取得するためのパラメータ
  */
message ServicePortInfo {
  string name = 1;								// サービスポート名
  repeated ServiceInterfaceInfo interfaces = 2;	// サービスインターフェース情報
}

/**
  * データポート情報を設定/取得するためのパラメータ
  */
message DataPortInfo {
  string name = 1;				// データポート名
  string type = 2;				// データポート型
  uint32 direction = 3;			// 方向(1:In, 2:Out)
}

/**
  * カテゴリ情報を設定/取得するためのパラメータ
  */
message CategoryInfo {
  string name = 1;				// カテゴリ名
  uint32 parent_id = 2;			// 親カテゴリのID
}

/**
  * コンポーネント情報を設定/取得するためのパラメータ
  */
message ComponentInfo {
  string name = 1;								// コンポーネント名
  string vendor = 2;							// ベンダ名
  uint32 language = 3;							// 実装言語(1:C++, 2:Python, 3:Java)
  string repository = 4;						// 対象コンポーネントが格納されているリポジトリのURI
  repeated CategoryInfo categories = 5;			// カテゴリ情報
  repeated DataPortInfo data_ports = 6;			// データポート情報
  repeated ServicePortInfo service_ports = 7;	// サービスポート情報
}

/**
  * コンポーネント情報を登録するためのパラメータ
  */
message ComponentRegistRequest {
  ComponentInfo component = 1;	// コンポーネント情報
}

/**
  * コンポーネント情報を検索するためのパラメータ
  */
message ComponentSearchRequest {
  repeated string category = 1;					// カテゴリに対する検索ワード
  bool category_or = 2;  						// カテゴリの検索条件.Falseの場合はAnd検索，Falseの場合はOr検索
  repeated string vendor = 3;					// ベンダに対する検索ワード
  bool vendor_or = 4;  							// ベンダの検索条件.Falseの場合はAnd検索，Falseの場合はOr検索
  repeated uint32 language = 5;					// 実装言語の検索条件．Or検索
  repeated string data_port_type = 6;			// データポートのデータ型に対する検索ワード
  bool data_port_type_or = 7;  					// データポートデータ型の検索条件.Falseの場合はAnd検索，Falseの場合はOr検索
  uint32 data_port_direction = 8;				// データポートの方向の検索条件．
  repeated string service_interface_type = 9;	// サービスインターフェースのインターフェース型に対する検索ワード
  bool service_interface_type_or = 10;			// サービスインターフェースのインターフェース型の検索条件.Falseの場合はAnd検索，Falseの場合はOr検索
  uint32 service_interface_direction = 11;		// サービスインターフェースの方向の検索条件．
}

/**
  * コンポーネント情報の検索結果を受け渡すためのパラメータ
  */
message ComponentSearchResponse {
  repeated ComponentInfo results = 1;			// コンポーネント情報の検索結果
}

/**
  * コンポーネント情報を更新するためのパラメータ
  */
message ComponentUpdateRequest {
  uint32 component_id = 1;	// 更新対象のコンポーネントのID
  string repository = 2;	// リポジトリ情報
}

/**
  * コンポーネント情報を削除するためのパラメータ
  */
message ComponentDeleteRequest {
  uint32 component_id = 1;	// 削除対象のコンポーネントのID
}

message StatusResponse {
  uint32 status = 1;	// 実行結果
}

/////
service ComponentDatabaseService {
  // コンポーネントを登録する
  rpc RegistComponent (ComponentRegistRequest) returns (StatusResponse) {}
  // コンポーネント情報を検索する
  rpc SearchComponent (ComponentSearchRequest) returns (ComponentSearchResponse) {}
  // コンポーネント情報を更新する
  rpc UpdateComponent (ComponentUpdateRequest) returns (StatusResponse) {}
  // コンポーネント情報を削除する
  rpc DeleteComponent (ComponentDeleteRequest) returns (StatusResponse) {}
}
